File: windows/AnywhereLLM.Core/UpdateCheck.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace AnywhereLLM.Core;

/// Pure logic for the self-updater: version comparison, GitHub release JSON
/// parsing, asset selection, and checksum-file parsing. No network or file I/O
/// here — that lives in the app's UpdateService. Mirrors Swift LLMCore/UpdateCheck.
public static class UpdateCheck
{
    public sealed record ReleaseAsset(string Name, string DownloadUrl, long Size);
    public sealed record ReleaseInfo(string Tag, IReadOnlyList<ReleaseAsset> Assets);

    /// True only if `latest` is strictly a higher semver than `current`. A leading
    /// "v" is stripped; unparseable input or an equal/older version returns false
    /// (this is the sole downgrade/reinstall guard).
    public static bool IsNewer(string current, string latest)
    {
        var c = Components(current);
        var l = Components(latest);
        if (c is null || l is null) return false;
        int n = Math.Max(c.Length, l.Length);
        for (int i = 0; i < n; i++)
        {
            int cv = i < c.Length ? c[i] : 0;
            int lv = i < l.Length ? l[i] : 0;
            if (lv != cv) return lv > cv;
        }
        return false; // equal
    }

    private static int[]? Components(string version)
    {
        var v = version.Trim();
        if (v.StartsWith("v", StringComparison.OrdinalIgnoreCase)) v = v[1..];
        if (v.Length == 0) return null;
        var parts = v.Split('.');
        var nums = new int[parts.Length];
        for (int i = 0; i < parts.Length; i++)
            if (!int.TryParse(parts[i], out nums[i])) return null;
        return nums;
    }

    /// Parse GitHub's /releases/latest response. Needs tag_name; assets missing
    /// name or browser_download_url are skipped. null on malformed JSON / no tag.
    public static ReleaseInfo? ParseLatestRelease(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return null;
            if (!root.TryGetProperty("tag_name", out var tagEl) || tagEl.ValueKind != JsonValueKind.String)
                return null;

            var assets = new List<ReleaseAsset>();
            if (root.TryGetProperty("assets", out var arr) && arr.ValueKind == JsonValueKind.Array)
            {
                foreach (var a in arr.EnumerateArray())
                {
                    if (a.TryGetProperty("name", out var n) && n.ValueKind == JsonValueKind.String &&
                        a.TryGetProperty("browser_download_url", out var u) && u.ValueKind == JsonValueKind.String)
                    {
                        long size = a.TryGetProperty("size", out var s) && s.ValueKind == JsonValueKind.Number
                            ? s.GetInt64() : 0;
                        assets.Add(new ReleaseAsset(n.GetString()!, u.GetString()!, size));
                    }
                }
            }
            return new ReleaseInfo(tagEl.GetString()!, assets);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    /// First asset whose name ends with the platform suffix (e.g. "-win-x64.zip"),
    /// so the MSI and the other platform's zip are ignored. null if none match.
    public static ReleaseAsset? PickAsset(IReadOnlyList<ReleaseAsset> assets, string suffix)
    {
        foreach (var a in assets)
            if (a.Name.EndsWith(suffix, StringComparison.OrdinalIgnoreCase)) return a;
        return null;
    }

    /// Parse a `<sha256>  <filename>` checksum file (sha256sum format; the optional
    /// "*" binary marker is tolerated). Non-conforming lines are ignored. Hashes
    /// are lowercased for case-insensitive comparison. Returns name -> hash.
    public static IReadOnlyDictionary<string, string> ParseChecksums(string text)
    {
        var map = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var raw in text.Split('\n'))
        {
            var m = Regex.Match(raw.Trim(), "^([0-9a-fA-F]{64})\\s+\\*?(.+)$");
            if (m.Success) map[m.Groups[2].Value.Trim()] = m.Groups[1].Value.ToLowerInvariant();
        }
        return map;
    }
}

File: Sources/LLMCore/UpdateCheck.swift
import Foundation

/// 자체 업데이터의 순수 로직: 버전 비교, GitHub 릴리즈 JSON 파싱, 에셋 선택,
/// 체크섬 파일 파싱. 네트워크·파일 I/O는 여기 없다 — 앱의 UpdateService 담당.
/// Windows AnywhereLLM.Core/UpdateCheck.cs와 동일 동작.

public struct ReleaseAsset: Equatable {
    public let name: String
    public let downloadURL: String
    public let size: Int
    public init(name: String, downloadURL: String, size: Int) {
        self.name = name
        self.downloadURL = downloadURL
        self.size = size
    }
}

public struct ReleaseInfo: Equatable {
    public let tag: String
    public let assets: [ReleaseAsset]
    public init(tag: String, assets: [ReleaseAsset]) {
        self.tag = tag
        self.assets = assets
    }
}

/// latest가 current보다 엄격히 높은 semver일 때만 true. 선행 "v" 제거. 파싱 불가/
/// 동일/더 낮음이면 false (유일한 다운그레이드·재설치 방지 게이트).
public func isNewer(current: String, latest: String) -> Bool {
    guard let c = versionComponents(current), let l = versionComponents(latest) else { return false }
    let n = max(c.count, l.count)
    for i in 0..<n {
        let cv = i < c.count ? c[i] : 0
        let lv = i < l.count ? l[i] : 0
        if lv != cv { return lv > cv }
    }
    return false // 동일
}

private func versionComponents(_ version: String) -> [Int]? {
    var v = version.trimmingCharacters(in: .whitespacesAndNewlines)
    if v.lowercased().hasPrefix("v") { v.removeFirst() }
    if v.isEmpty { return nil }
    var nums: [Int] = []
    // omittingEmptySubsequences: false — "0..1" 같은 빈 성분은 파싱 실패로 nil (C# Split과 동일).
    for part in v.split(separator: ".", omittingEmptySubsequences: false) {
        guard let num = Int(part) else { return nil }
        nums.append(num)
    }
    return nums
}

/// GitHub /releases/latest 응답 파싱. tag_name 필수; name/browser_download_url 없는
/// 에셋은 건너뜀. JSON 불량/태그 없음이면 nil.
public func parseLatestRelease(_ json: Data) -> ReleaseInfo? {
    guard let obj = (try? JSONSerialization.jsonObject(with: json)) as? [String: Any],
          let tag = obj["tag_name"] as? String else { return nil }
    var assets: [ReleaseAsset] = []
    if let arr = obj["assets"] as? [[String: Any]] {
        for a in arr {
            guard let name = a["name"] as? String,
                  let url = a["browser_download_url"] as? String else { continue }
            let size = (a["size"] as? Int) ?? 0
            assets.append(ReleaseAsset(name: name, downloadURL: url, size: size))
        }
    }
    return ReleaseInfo(tag: tag, assets: assets)
}

/// 이름이 플랫폼 접미사로 끝나는 첫 에셋(예: "-macos.zip") — MSI와 다른 플랫폼 zip은
/// 무시. 매치 없으면 nil.
public func pickAsset(_ assets: [ReleaseAsset], suffix: String) -> ReleaseAsset? {
    let s = suffix.lowercased()
    return assets.first { $0.name.lowercased().hasSuffix(s) }
}

/// `<sha256>  <filename>` 체크섬 파일 파싱(sha256sum 형식; "*" 바이너리 마커 허용).
/// 형식 불일치 줄 무시. 해시는 소문자로. name -> hash 반환.
public func parseChecksums(_ text: String) -> [String: String] {
    var map: [String: String] = [:]
    let regex = try! NSRegularExpression(pattern: "^([0-9a-fA-F]{64})\\s+\\*?(.+)$")
    for rawLine in text.split(separator: "\n", omittingEmptySubsequences: false) {
        let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines) // CRLF의 \r 포함 제거
        let range = NSRange(line.startIndex..., in: line)
        guard let m = regex.firstMatch(in: line, range: range),
              let hashR = Range(m.range(at: 1), in: line),
              let nameR = Range(m.range(at: 2), in: line) else { continue }
        let name = line[nameR].trimmingCharacters(in: .whitespaces)
        map[name] = line[hashR].lowercased()
    }
    return map
}
